ch3/exercise3-3/surface.go

```diff
@@ -14,41 +15,79 @@ const (
 	angle         = math.Pi / 6         // angle of x, y axes (=30°)
 )
 
+const (
+	peak   = 1
+	valley = 2
+)
+
 var sin30, cos30 = math.Sin(angle), math.Cos(angle) // sin(30°), cos(30°)
 
 func main() {
-	fmt.Printf("<svg xmlns='http://www.w3.org/2000/svg' "+
+	// outputting into file so I can view the result in my code editor with svg preview
+	file, fileErr := os.Create("surface.svg")
+	if fileErr != nil {
+		fmt.Println(fileErr)
+		return
+	}
+	fmt.Fprintf(file, "<svg xmlns='http://www.w3.org/2000/svg' "+
 		"style='stroke: grey; fill: white; stroke-width: 0.7' "+
 		"width='%d' height='%d'>", width, height)
+	fmt.Fprintf(file, "<style type=\"text/css\">")
+	fmt.Fprintf(file, ".blue {fill: #0000FF; }\n .red {fill:#FF0000;}")
+	fmt.Fprintf(file, "</style>")
+
 	for i := 0; i < cells; i++ {
 		for j := 0; j < cells; j++ {
-			ax, ay := corner(i+1, j)
-			bx, by := corner(i, j)
-			cx, cy := corner(i, j+1)
-			dx, dy := corner(i+1, j+1)
-			fmt.Printf("<polygon points='%g,%g %g,%g %g,%g %g,%g'/>\n",
-				ax, ay, bx, by, cx, cy, dx, dy)
+			ax, ay, type1 := corner(i+1, j)
+			bx, by, type2 := corner(i, j)
+			cx, cy, type3 := corner(i, j+1)
+			dx, dy, type4 := corner(i+1, j+1)
+
+			var current string
+
+			if type1 == valley || type2 == valley || type3 == valley || type4 == valley {
+				current = "blue"
+			} else if type1 == peak || type2 == peak || type3 == peak || type4 == peak {
+				current = "red"
+			} else {
+				current = "grey"
+			}
+
+			fmt.Fprintf(file, "<polygon points='%g, %g %g,%g %g,%g %g,%g'" + " class=\"%s\"/>\n",
+				ax, ay, bx, by, cx, cy, dx, dy, current)
+
 		}
 	}
-	fmt.Println("</svg>")
+	fmt.Println(file, "</svg>")
 }
 
-func corner(i, j int) (float64, float64) {
+func corner(i, j int) (float64, float64, int) {
 	// Find point (x,y) at corner of cell (i,j).
 	x := xyrange * (float64(i)/cells - 0.5)
 	y := xyrange * (float64(j)/cells - 0.5)
 
 	// Compute surface height z.
-	z := f(x, y)
+	z, kind := f(x, y)
 
 	// Project (x,y,z) isometrically onto 2-D SVG canvas (sx,sy).
 	sx := width/2 + (x-y)*cos30*xyscale
 	sy := height/2 + (x+y)*sin30*xyscale - z*zscale
-	return sx, sy
+	return sx, sy, kind
 }
 
-func f(x, y float64) float64 {
+// finding the 2nd derivative of f(x) where f(x)= sin(x)/x
+// math taken from https://xingdl2007.gitbooks.io/gopl-soljutions/content/chapter-3-basic-data-types.html
+// because I don't remember calc too much
+
+func f(x, y float64) (float64, int) {
+	d := math.Hypot(x, y)
 	r := math.Hypot(x, y) // distance from (0,0)
-	return math.Sin(r) / r
+	kind := 0
+	if math.Abs(d-math.Tan(d)) < 3 {
+		kind = peak
+		if 2*(math.Sin(d)-d*math.Cos(d))-d*d*math.Sin(d) > 0 {
+			kind = valley
+		}
+	}
+	return math.Sin(r) / r, kind
```

Comment:
The book's phrasing was a bit ambiguous I suppose.  I interpreted it as asked to color based on height, whereas this approach colors based on local min/maxima (1st derivative), which has little to do with height, but has everything to do with peaks and valleys.  I'd imagined something more like:

![out](https://user-images.githubusercontent.com/187314/90166100-895d4480-dd4e-11ea-8b8b-4999cfec3055.png)

The above image doesn't require calculus, but it does require determining the max/min z values in a precompute, maybe grappling with infinite values and calculating a sliding color scale, which I think is interesting from a programming standpoint and much more applicable to the graphing work you'll be doing for `tsh bench`.

Up to you if you want to adjust. I recognize the schedule is pretty aggressive and don't want you to get bogged down.